### DISCUSSION TITLE
Timescaledb (Postgresql) support for time_bucket_gapfill

### DISCUSSION THREAD

**Discussion by ajspera** [grafana/grafana#37003] [Data Sources]
TimescaleDB v1.2 has introduced a time bucket gap-filling function. This should be implemented for grafana's query generator.

I was experimenting with TimescaleDB today and noticed gap filling wasn't being used in the query and without the null values Grafana doesn't fill intervals itself. Coincidentally I noticed TimescaleDB had released that feature 2 days ago, so I figured I'd put a feature request in for the option to use it in the generator UI.

`time_bucket_gapfill('1 day', time, '2017-09-01', '2017-10-01') AS date`
vs
`time_bucket('1 day', time) AS date`

**Comment by marefr** [comment]
@ajspera Just to be clear. If you use the raw query editor you can use `time_bucket_gapfill` now given you're Timescale DB instance supports it. It's only in the graphical query builder it's not yet  and why this feature request is opened, correct?

**Comment by ajspera** [comment]
Yeah, this is just for the query builder gui. I'm using it manually in the raw editor.

**Comment by aocenas** [comment]
@ajspera I think you can already do what you want in the query builder:

<img src="https://user-images.githubusercontent.com/1014802/53739444-10d43200-3e92-11e9-9243-9f8b3ec938e8.gif" height="300">

The query itself does not change because Grafana is doing the filling in [backend code](https://github.com/grafana/grafana/blob/master/pkg/tsdb/sql_engine.go#L319) before the results are returned but there is a macro that is sent to backend which controls the fill behaviour `SELECT $__timeGroupAlias("time",5s,0) ...` (this will fill the intervals with `0`s)

Changing this to use `time_bucket_gapfill` could have some perf benefit (would need to benchmark that) on the other hand it would add additional complexity to the query engine as I think right now the gap filling code is same for both Postgres and MySQL.

**Comment by marefr** [comment]
Good points @aocenas :+1:

**Comment by nonujigu** [comment]
Great, setting the variable that follows the $__interval variable to NULL was exactly the feature that I was looking for. Unfortunately it is not working anymore if you go back to text edit mode. Once your queries have a certain complexity in text edit mode you cannot go back to the automatic query builder anymore and Grafana will just interpolate big time ranges that have no data. There seems to be no way to set that variable in text edit mode.

Another problem I have with the $__interval variable is that it never gets smaller than '1s' although hardcoding it with something like '0.01s' is possible. The actual problem is that the min time interval parameter in the data sources configuration tab is set to ‘1s’ and I am unable to set it to a lower value. According to the [this official Grafana Doc](https://grafana.com/docs/features/datasources/postgres/#min-time-interval) it should be possible to set it to a lower value by using the 'ms' time identifier but Grafana is just not swallowing that.
![a7d84ae9a2c76c5becbf295d6ba7858a442badfe](https://user-images.githubusercontent.com/49524829/61171196-83f26100-a574-11e9-9665-327da7d8335b.png)

**Comment by nuKs** [comment]
@aocenas solutions work like a charm.

A few notes:
- For some reason, $__timeGroupAlias doesn't work with SQL windowing for me.
- time_bucket_gapfill turns value to NULL by default and thus has required me to use coalesce to adds up for the 0 feature.

Thus my final code,
```SQL
SELECT
  time_bucket_gapfill('$__interval', "timestamp") as time,
  max(count(x)) over (PARTITION BY 1) - coalesce(count(x), 0) AS "x"
FROM
  accelerometer
WHERE
  $__timeFilter("timestamp")
GROUP BY 1
ORDER BY 1
```

**Comment by gabor** [comment]
based on the comments it seems a working solution was proposed. i see there are (interval never gets smaller than `1s` etc.), for each of those i recommend creating new, separate github issues. i am closing this discussion as the original problem has a solution. thanks all!
